### PR TITLE
[x11 layering] Fix bug with multiple transients

### PR DESCRIPTION
**Pull request by elParaguayo**
Google Earth identified a couple of issues with the layering code:
- The `transient_for` property for some windows returned the ID of a window that was not being managed by qtile. As a result, the parent window would not have been stacked and placing the new window above this would therefore put the new window in the wrong stack position.
- Transient windows were excluded from the list of windows in different layers and so, if an app has multiple transient windows, a new window would not be stacked after the last transient.

Draft for now as I'd like some feedback.

### CODE REVIEW

File: libqtile/backend/x11/window.py [hunk 1]
```diff
@@ -1044,14 +1044,15 @@ def change_layer(self, up=True, top_bottom=False):
         windows.sort(key=lambda w: stack.index(w[0].wid))
 
         # Get lists of windows on lower, higher or same "layer" as window
-        lower = [w[0].wid for w in windows if w[1] is None and w[2] > layering]
-        higher = [w[0].wid for w in windows if w[1] is None and w[2] < layering]
-        same = [w[0].wid for w in windows if w[1] is None and w[2] == layering]
```
Comment (elParaguayo) [inline libqtile/backend/x11/window.py:1049]: Is there any reason why we should exclude transient windows when calculating the list of windows in lower, same and higher layers? I can't think of any.

If we don't need to exclude them then I think we can remove the transient field from the windows tuple.

File: libqtile/backend/x11/window.py [hunk 2]
```diff
@@ -1044,14 +1044,15 @@ def change_layer(self, up=True, top_bottom=False):
...
+        lower = [w[0].wid for w in windows if w[2] > layering]
+        higher = [w[0].wid for w in windows if w[2] < layering]
+        same = [w[0].wid for w in windows if w[2] == layering]
 
         # We now need to identify the new position in the stack
 
         # If the window has a parent, the window should just be put above it
-        if parent:
+        # If the parent isn't being managed by qtile then it may not be stacked correctly
+        if parent and parent in self.qtile.windows_map:
```
Comment (elParaguayo) [inline libqtile/backend/x11/window.py:1055]: I've no idea why Google Earth says that some dialogs are transient windows for a window that we're not managing (and not displaying). Is there a better way to handle that scenario>

Comment (m-col) [reply libqtile/backend/x11/window.py:1055]: If we aren't managing the parent window, could we ignore the transient hint and treat it as a normal window?

File: libqtile/backend/x11/window.py [hunk 3]
```diff
@@ -1044,15 +1044,32 @@ def change_layer(self, up=True, top_bottom=False):
...
+            # If the window is modal then it should be placed above every other window that is in that window group
+            # e.g. the parent of the dialog and any other window that is also transient for the same parent.
+            if "_NET_WM_STATE_MODAL" in self.window.get_net_wm_state():
+                window_group = [parent]
+                window_group.extend(
+                    k
+                    for k, v in self.qtile.windows_map.items()
+                    if v.window.get_wm_transient_for() == parent
+                )
+                window_group.sort(key=lambda w: stack.index(w))
```
Comment (m-col) [inline libqtile/backend/x11/window.py:1065]:
```suggestion
                window_group.sort(key=stack.index)
```

File: libqtile/backend/x11/window.py (hunk 1 above)
Comment (m-col) [reply libqtile/backend/x11/window.py:1049]: The only thing that matters is they are above their parent, but that doesn't require excluding them here does it?

File: libqtile/backend/x11/window.py (hunk 2 above)
Comment (elParaguayo) [reply libqtile/backend/x11/window.py:1055]: That's essentially what that line does. i.e. if we're not managing it, this condition isn't met and we'll fall back to normal rules.

**Comment by elParaguayo** [comment]
I probably need to add something that moves transients back about the parent if the parent is brought to the front.

**Review by m-col — APPROVED** [review]

**Pull request merged**